Added self healing proposal (#2)
ShubhamRwt/proposals: 087-self-healing-using-cruise-control.md

# Self Healing using Cruise Control

> This proposal is about integrating the [self-healing](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) feature of Cruise Control into Strimzi operator.

Reviewer: ```suggestion
This proposal is about integrating the self-healing feature of Cruise Control into the Strimzi operator.
```

The link points just to the self-healing configuration page which doesn't help to understand what self-healing is, so I would just remove it.

Reviewer 2: Yes, or link to the part of the docs that cover the anomaly detector and how it integrates.

Reviewer 2: _Side Note_: If this proposal is accepted that there might be scope for writing more detailed self-healing docs in the upstream.

> The self-healing feature of cruise control allows us to detect the anomalies using its anomaly detector mechanism and then based on the anomaly, cruise control can generate an optimization proposal that fixes the anomaly and move the cluster into healthy state again.

Reviewer: I think you should rephrase this, because technically it's not self-healing detecting anomalies. Self-healing provides the support to fix anomalies which are automatically detected by the anomaly detector. So while the two things are related, they are different. Take into account that anomaly detection is already usable in Strimzi. You can set your notifier as well but you don't have a way to fix automatically.

> ## Current situation

Reviewer: I would also explain that currently, Strimzi has the support to enable anomalies detection and a corresponding notifier for that but what's missing is an automatic way to fix those anomalies.

> When the users are working with Kafka clusters, it is very much possible that the cluster can move to unhealthy state sure to some broker failure, disk failure or some other issue.

Reviewer: ```suggestion
When the users are working with Kafka clusters, it is very much possible that the cluster can move to unhealthy state due to some broker failure, disk failure or some other issues.
```

> When the users are working with Kafka clusters, it is very much possible that the cluster can move to unhealthy state sure to some broker failure, disk failure or some other issue.

Reviewer 2: ```suggestion
During normal operations it common for Kafka clusters to become unbalanced (through partition key skew, uneven partition placement etc) or for problems to occur with disks or other underlying hardware which makes the cluster unhealthy.  
```

> Currently, if we encounter any such scenario we need to fix these issues manually i.e. if there is some broker failure then we might move the partition replicas form that corrupted broker to some other healthy broker using the  `kafka-reassign-partitions.sh` tool.

Reviewer: we can use the `KafkaRebalance` custom resource to remove the broker. The reassignment tool should not be the way to go anymore.

Reviewer 2: Yes I would recommend Strimzi's native support for removing brokers with CC and link to the docs.

> While with small sized cluster it's easier to fix things manually but when it comes to very big cluster it quite hard to do everything manually since it would take a lot of time to fix all of those issues and then getting the cluster in healthy state.

Reviewer: ```suggestion
With small sized cluster, it's easy to fix things manually but when it comes to very big cluste, it's quite hard to do everything manually, since it would take a lot of time to fix all of those issues and then getting the cluster in healthy state.
```

> While with small sized cluster it's easier to fix things manually but when it comes to very big cluster it quite hard to do everything manually since it would take a lot of time to fix all of those issues and then getting the cluster in healthy state.

Reviewer 2: ```suggestion
While with small sized cluster it is feasible to fix things manually, when it comes to larger ones it can be very time consuming or just not feasible to fix all the issue on your own.
```

> Also, its very much possible while doing thing manually that we might lose some data or maybe mess up the cluster completely.

Reviewer 2: ```suggestion
Also, it is possible that while doing thing manually, that we might lose data or cause other unforeseen issues with the Kafka Cluster.
```

## Motivation

With the help of the self-healing feature we can resolve the issues like disk failure, broker failure and other issues automatically in the cluster.

> Cruise Control treats these issues as `anomalies`. Cruise will be using the anomaly detector mechanism to detect any anomalies.

Reviewer: ```suggestion
Cruise Control treats these issues as `anomalies` which are detected by using an anomaly detector mechanism.
```

> Cruise Control treats these issues as `anomalies`. Cruise will be using the anomaly detector mechanism to detect any anomalies.

Reviewer 2: ```suggestion
Cruise Control treats these issues as "anomalies", the detection of which is the responsibility of the [anomaly detector mechanism](<link to docs>).
```

> If the anomalies are present then a response will be passed on to the operator by the self-healing notifier and then if self is healing configuration is enabled, the cluster will be healed automatically by Cruise Control without any user intervention required.

Reviewer: This is not a motivation. This is a short description about what's going to happen.

Reviewer 2: I agree with Paolo, you need to describe how this is currently disabled in Strimzi, why that was the case in the past and **why** we need to integrate this into Strimzi now.

## Proposal

> This proposal is about introducing the self-healing feature of Cruise Control into the strimzi operator.

Reviewer 2: ```suggestion
This proposal is about introducing the self-healing feature of Cruise Control into the Strimzi Operator.
```

> This feature would allow the cluster to be healed automatically if there are some anomalies detected.

Reviewer 2: I think this sentence is redundant, you have established what self healing does in the sections above. You don't need to say it again here.

Anomalies are basically issues which can arise in the cluster due to various reason like disk failure, broker failure, topic related issues etc.

> Cruise control detects these anomalies using the Anomaly detector and then heals the cluster.

Reviewer: I think you already stated the above four lines too many times in the previous paragraphs.

Reviewer 2: Agreed, you have said this above.

### Enabling self-healing in Kafka resource

> To enable the self-healing feature, the user should set the `self.healing.enabled` configuration to `true` while passing configurations regarding Cruise Control in the Kafka custom resource.

Reviewer 2: ```suggestion
To enable the self-healing feature, the user will set the `self.healing.enabled` configuration (in the `spec.kafka.cruiseControl.config`) to `true` in the Kafka custom resource.
```

> Self-healing should be disabled by default.

Reviewer 2: ```suggestion
As currently, the self-healing feature will be disabled by default.
```

> In the following example, the user can decide to enable the self-healing feature of cruise control in the Kafka custom resource:

Reviewer 2: I don't think you need this description, the comments in the example code say enough. I would delete it.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true  # false by default 
```

> Once self-healing is enabled, that would mean that it will now detect all the anomalies like disk failure, broker failure, topic anomaly etc. and then heal the cluster automatically without any user interaction.

Reviewer: This is not technically true. The anomalies detection is not something you can disable in Cruise Control. It's always enabled. The detectors are always running. The point is that the notifier is, by default, a NoopNotifier. It means that even if an anomaly is detected, the notification is get lost and the anomaly is always "IGNOREd".
If you just enable self-healing without using a specific notifier by setting `anomaly.notifier.class` you will get no anomaly fixed anyway.

Reviewer 2: I think the upstream docs do a bad job of explaining all this. I think it is worth putting together a short explanation of the anomaly detection to self healing flow in the introduction sections so readers have the context. It is important that the reader understands this.

In case the user wants to enable self-healing for some specific anomalies then they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.

> They also have the flexibility to configure properties like the rate at which anomalies should be detected, the time before a fix should happen etc. and many [other configuration properties](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) just like these.

Reviewer: Some configuration parameters are part of the anomaly detection not the self-healing itself and they are here https://github.com/linkedin/cruise-control/wiki/Configurations#anomalydetector-configurations
I think you should try to make a good distinction through this proposal about what's anomaly detection and what's self-healing. Again, related each other but two separate things.

Reviewer 2: +1, another argument for explaining the process of anomaly detection to triggering self healing and how they are both configured.

### Interaction between the Operator and Cruise Control

> A two-way interactions should be made between the operator and cruise control such that the user get to turn of self-healing whenever they desire and also so that the operator can receive updates on what anomalies are being fixed or detected.

Reviewer 2: ```suggestion
There will be a two-way interactions between the Strimzi Operator and Cruise Control, such that the user has the option to pause self-healing and also so that the operator can receive updates on what anomalies are being fixed or detected.
```

Reviewer 2: Does the CO actually care what anomalies are being detected? Does it use that information or is it just for the human user?

> To establish this two-way communication, we are going to leverage the `Kubernetes custom resource` which can be used by both cruise control and the operator when running on Kubernetes.

Reviewer 2: ```suggestion
To establish this two-way communication, we are going to leverage a new custom resource, which can be used by both Cruise Control and the Strimzi Operator.
```

Reviewer 2: You can only run the operator on K8s, so there is no need to mention that explicitly.

Reviewer 2: Also, you might go over this later, but do you need to say what this new CR will be?

#### Strimzi Notifier and Cruise Control role

> Cruise Control uses anomaly detector classes to detect the anomalies and then based on those anomalies it provides the ability to alert the users that some anomaly was detected and self-healing is going to fix it.

Reviewer: I would use "notify" and not the "alert" word which is part of the `SelfHealingNotifier` class implementation.

> Cruise Control uses anomaly detector classes to detect the anomalies and then based on those anomalies it provides the ability to alert the users that some anomaly was detected and self-healing is going to fix it.

Reviewer 2: ```suggestion
Cruise Control uses specific detector classes to identify relevant anomalies and then, based on those anomalies, notify the users (via the [notifier system](<link-to-docs>)) that an anomaly was detected and that self-healing is going to fix it.
```

> These alerts are triggered by the `SelfHealingNotifier` class in cruise control.

Reviewer: That's not again technically true. You can put the notifier you want through the `anomaly.notifier.class` configuration. The class needs to implement the `AnomalyNotifier` interface.
The `SelfHealingNotifier` class is just a good base class for building your own custom notifier but it's not mandatory. You can just develop your own custom notifier by implementing the `AnomalyNotifier` interface and then getting the notification and doing whatever you want with it.

> But just an alert is not good if we want to have the interaction going between the operator and cruise-control, therefore we would need a new `custom` notifier which can be used instead of the inbuilt `SelfHealingNotifier` to set up a way for having an interaction between cruise control and operator

Reviewer 2: ```suggestion
But a notification alone is not particularly useful if we want to have two-way communication between the operator and Cruise Control. Therefore, we need a new notifier implementation, which can be used instead of the inbuilt `SelfHealingNotifier`, to facilitate that communication.
```

> Strimzi Notifier will be a new `custom` notifier which should be used in place of the `SelfHealingNotifier` which is used by cruise control.

Reviewer: Again "`SelfHealingNotifier` which is used by cruise control." not technically true ;-)

> To implement it, we will be extending the `SelfHealingNotifier` class and overriding the methods that we need.

Reviewer: It's too generic ... which are the methods that we need? I think you should have a short intro about the `AnomalyNotifier` interface so what the `SelfHealingNotifier` provides.
I also think that the readers also need to understand a bit about the notifier and that it makes the decision about an anomaly being fixed or ignored ... it's not just notifying.
Also, does the `StrimziNotifier` need any configuration for its purpose?

Reviewer 2: +1, again you need a more detailed introduction showing the flow from detection > notification > self-healing. Writing that out will be vital context for the reader, good reference material for you and could be contributed upstream to the CC docs.

> The new notifier can be easily used by adding another configuration property in the cruise control section of the Kafka custom resource called `anomaly.notifier.class`.

Reviewer: At this point you should have already explained what is the `anomaly.notifier.class` property as mentioned more times above.

Here is an example:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true  # false by default 
      anomaly.notifier.class: io.strimzi.kafka.notifier.StrimziNotifier
```

> The `StrimziNotifier` class is going to reside in a separate module in the operator repository called the `strimzi-notifier` so that we can later build a jar w.r.t it and then include that jar in the cruise-control image. 

Reviewer: it's technically the Kafka image which we use to package Cruise Control as well.

> This is required since we need to run the notifier as part of the cruise control image.

Reviewer: ```suggestion
This is required since we need to run the notifier as part of Cruise Control.
```

> We would also require to add some fabric8 client dependencies to the cruise control image since we are going to use them for creating the anomaly custom resource.

Reviewer 2: Might be an interesting process. You may need to build a fat-jar with the fabric8 libs in as I don't _think_ CC shares a class-path with Strimzi.

Reviewer: The CC part (a.k.a. the `StrimziNotifier`) is going to be part of the Kafka image where we have a dedicated CC folder with related deps. Adding fabric8 there is a pain point in the POC so I think you raised a very good point about having a fat-jar to avoid this. Even because fabric8 itself is bringing some jackson related deps which also CC has so there is the risk of conflicts there.

With the new `Strimzi notifier` we would be overriding how alerting currently works in CC.

> Everytime cruise control detects an anomaly, our new notifier is going to create a `KafkaAnomaly` custom resource which should contain information about the anomaly like anomaly-id, anomaly-type etc.

Reviewer 2: ```suggestion
Every time Cruise Control detects an anomaly, our new notifier will create a `KafkaAnomaly` custom resource which should contain information about the anomaly like anomaly-id, anomaly-type etc.
```
You introduce the new CR in the section above this but don't give it a name. So might be worth either removing the mention of it above or adding some details up there.

Also, can you please make sure you use consistent capitalisation, Cruise Control is a proper name so always needs to be capitalised.

Example:
```yaml
- apiVersion: kafka.strimzi.io/v1beta2

>   kind: Anomaly

Reviewer: ```suggestion
  kind: KafkaAnomaly
```

  metadata:
    creationTimestamp: "2024-11-12T12:01:58Z"
    generation: 1

>     name: my-cluster-anomaly

Reviewer: Is it going to be the name of the custom resource? Just <cluster name> followed by "anomaly"? Ho to distinguish different anomalies happening in the same cluster? What when the same anomaly, like a topic anomaly, can have different flavours (i.e replication factor error, too big partition size, ...).

Reviewer 2: +1 names need to be unique so you definitely need the name of the Kafka cluster, type of anomoly and some unique identifier (you will need to watch the length of the `anomalyID` uuid as it might breach the name length limit when combined with the cluster name and anomaly type). 

Something like: 
```suggestion
    name: my-cluster--disk-failure--abc7b448f936
 ```

    namespace: myproject
    resourceVersion: "5704"
    uid: 95cc2621-052c-440e-a553-f14d003669be
  spec:
    anomalyId: a69c2d7a-ff53-40db-8cd0-abc7b448f936
    anomalyType: DISK_FAILURE

>     anomalyProgress: FixPending

Reviewer: this would be part of the status, what does it represent here?

  status:
#......
```

> the `anomalyId` represents the id assigned to the anomaly, or you can also say it is the user task id.

Reviewer: being a user task id is an internal details in the Cruise Control anomaly detection implementation so I don't think it's useful to mention it here.

Reviewer 2: +1, remove

the `anomalyType` represents the type of anomaly that is detected.
the `anomalyProgress` represents the progress of the anomaly.

> It can have different values like `FixPending`, `FixComplete` or `FixFailed`.

Reviewer: See my above comment, the progress should be part of the status.

> This Anomaly custom resource is going to be the source of interaction between both operator and cruise control.

Reviewer 2: ```suggestion
This `KafkaAnomaly` custom resource is going to be the source of interaction between both the operator and Cruise Control.
```

> So if self-healing is enabled and some anomaly is detected, a Kafka Anomaly resource will be deployed in the custer.

Reviewer 2: ```suggestion
So if self-healing is enabled and some anomaly is detected, a `KafkaAnomaly` resource will be deployed in the Kubernetes cluster, in the same namespace as the relevant `Kafka` resource.
```

> The operator is going to watch the `KafkaAnomaly` resource with the help of Kubernetes watcher.

Reviewer 2: The people reviewing this proposal are K8s operator developers, they will know this so you don't need to spell it out. 

However, you do need to say what the cluster will do when it reconciles these CRs. If a user changes something in the `KafkaAnomaly` CR what will happen. Is it a 2-way sync, 1-way?

> Since we need to create `KafkaAnomaly` resource through the notifier, we would need to add some `ClusterRole` and `ClusterRoleBindings` w.r.t to the anomaly custom resource which we are going to be created.

Reviewer 2: ```suggestion
We will need to create `KafkaAnomaly` resource via the custom Cruise Control notifier implementation. 
Therefore, we will need to add appropriate `ClusterRole` and `ClusterRoleBindings` to allow this.
```

These role and role binding yaml's should be part of the operator installation so that the user have them deployed when they are installing the operator.

> That would provide them ease to easily enable self-healing whenever they require it.

Reviewer 2: This sentence is implied by the one above. I would remove it.

> #### Operator role

Reviewer 2: "Role" is an overloaded term in K8s. I think a better title would be "Reconciliation process in the Strimzi Cluster operator" or similar.

> Once the `KafkaAnomaly` resource is created then the operator can watch this resource by the help of the Kubernetes watcher.

Reviewer: It's not clear what do you mean just by "Kubernetes watcher". Where is it going to live?
I think you are missing some details about how watching a `KafkaAnomaly` has an impact in terms of implementation on the operator. Which part of the operator is going to "watch" for `KafkaAnomaly` resources? Is it new? Is it anything already existing?

> If an anomaly is detected then the operator work is to update the `.status` field of the anomaly resource with the current progress of the anomaly i.e. whether it is fixed or not

Reviewer 2: ```suggestion
If an anomaly is detected then the operator's job is to update the `.status` field of the anomaly resource with the current progress of the anomaly i.e. whether it is fixed or not
```

> When an anomaly is detected, Cruise assigns a state to the anomaly i.e.DETECTED. In the same way various states are assigned to anomaly as per the action taken on it.

Reviewer 2: ```suggestion
When an anomaly is detected, Cruise Control assigns a state to the anomaly (e.g. `DETECTED`). In the same way various states are assigned to anomaly as per the action taken on it.
```

```

This query should return us the state of the anomaly. Here is an example result of the query:
```shell
AnomalyDetectorState: {selfHealingEnabled:[BROKER_FAILURE, DISK_FAILURE, GOAL_VIOLATION, METRIC_ANOMALY, TOPIC_ANOMALY, MAINTENANCE_EVENT], selfHealingDisabled:[], selfHealingEnabledRatio:{BROKER_FAILURE=1.0, DISK_FAILURE=1.0, GOAL_VIOLATION=1.0, METRIC_ANOMALY=1.0, TOPIC_ANOMALY=1.0, MAINTENANCE_EVENT=1.0}, recentGoalViolations:[], recentBrokerFailures:[{anomalyId=37d4c9be-675a-4124-bf35-83aba3410b3c, failedBrokersByTimeMs={3=1731574274942}, detectionDate=2024-11-14T08:51:14Z, statusUpdateDate=2024-11-14T08:51:14Z, status=FIX_STARTED}], recentMetricAnomalies:[], recentDiskFailures:[], recentTopicAnomalies:[], recentMaintenanceEvents:[], metrics:{meanTimeBetweenAnomalies:{GOAL_VIOLATION:0.00 milliseconds, BROKER_FAILURE:22.19 milliseconds, METRIC_ANOMALY:0.00 milliseconds, DISK_FAILURE:0.00 milliseconds, TOPIC_ANOMALY:0.00 milliseconds}, meanTimeToStartFix:0.00 milliseconds, numSelfHealingStarted:0, numSelfHealingFailedToStart:0, ongoingAnomalyDuration=6.91 seconds}, ongoingSelfHealingAnomaly:None, balancednessScore:100.000}
```

>  We can then update the `.status` field of the `KafkaAnomaly` resource with the current status of the resource and other details if required.

Reviewer: the anomaly detector state is a cache of all the latest anomalies, you need to do some filtering here. You should explain how to get the info about the anomaly you are interested in.

> The operator is also responsible to make sure that once the anomaly is fixed, the resource with respect to that particular anomaly should be removed. 

Reviewer: This is something still under discussion.

To know if the anomaly is resolved or not, we again need to poll the `/kafkacruisecontrol/state` endpoint but this time instead of polling the `anomaly_detector` substate we are going to poll the `executor` substate.

> The `executor` substate tell us if the action for fixing the anomaly is complete or not.

Reviewer: Actually it doesn't say it. This is what we are totally missing. We are thinking about a workaround to realize that the anomaly was fixed but no one is telling us.

Here is an example request:
```sh
curl -X GET "http://my-cluster-cruise-control:9090/kafkacruisecontrol/state?verbose &json=true&substates=executor"
```

> and here is the result w.r.t to this query:

Reviewer 2: ```suggestion
and here is the result of this query:
```
Can you not use "w.r.t", "with respect to" is often just a long way of say "to" or "of" or similar. Also not all readers will know what the acronym means.

Reviewer: me! me! ... I always forget what w.r.t. means :-D

```sh
ExecutorState: {state: NO_TASK_IN_PROGRESS}
```
 if it says `NO_TASK_IN_PROGRESS`, that means that either the task is finished or there was no task started, if there is some task in progress, then you should get some output like this
```sh
ExecutorState: {state: INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS, pending(7)/in-progress(0)/aborting(0)/finished(0)/total(7) inter-broker partition movements, completed(0)/total(0) bytes in MBs: 0.00%, max/min/avg concurrent inter-broker partition movements per-broker: 5/5/5.00, triggeredSelfHealingTaskId: 37d4c9be-675a-4124-bf35-83aba3410b3c, triggeredTaskReason: Self healing for BROKER_FAILURE: {Fixable broker failures detected: {Broker 3 failed at 2024-11-14T08:51:14Z}}, recentlyRemovedBrokers: [3]}
```

> The `triggeredSelfHealingTaskId` mentioned in the output is similar to the `anomalyId` of anomaly that is being fixed. 

Reviewer 2: "Similar"? Is it the same or not?

> To make sure that when the anomaly is fixed the corresponding resource gets deleted the operator will set the `.spec.anomalyProgess` to `FixComplete` once the fix is completed and polling the `Excecutor` state says that there is no task in progress.

Reviewer: what if the executor doesn't return NO_TASK_IN_PROGRESS but returns a new task running? What about the anomaly status you were looking for?

`StrimziNotifier` will also have a Kubernetes watcher which will be checking if the `anomalyProgress` field changed or not.

> If it changed to `FixComplete` then it means that fix was complete and the resource can then be deleted by the notifier or if it says `FixFailed` then we should log the error regarding the reason why it failed(i.e a broker failure that cannot be fixed or some other issue).

Reviewer: Still to discuss if the resource has to be deleted, anyway not sure it would be the `StrimziNotifier`.

Reviewer 2: Yes, as discussed the whole point of these `KafkaAnomaly` CRs (AFAIUI) is to inform human users that anamalies were detected and fixed. So you need to have them hang around for a configurable time so human users can use them in debugging issues.

Reviewer 2: This also might be the biggest area of push-back. How many of these CRs are we going to create, at what rate, are they going to put burden on the CO. You will need answers to those questions.

Reviewer: This is strictly related to how the anomaly detection mechanism assign a new id even when the "anomaly" is the same. It will drive the decision of the naming for each `KafkaAnomaly` instance as Tom already mentioned before. It's something to think through more ... I haven't reached a good compromise as well :-(

> ### Execution Flow

Reviewer: I think the flow is not really clear.
You should try to create separate subparagraphs for it:

- explain the process of the `KafkaAnomaly` resource creation.
- the process for the operator to get the fix status and updating the resource.

Reviewer 2: +1, this section needs to be **very** clear. I think numbered steps with a flow diagram (with reference to the numbered steps) would be a big improvement.

- The user would first enable self-healing in cruise control configuration by setting the `self.healing.enabled` field as `true` and also overriding the `anomlay.notifier.class` to use the new `Strimzi` notifier. 

> - Once the user has enabled self-healing, now the cluster is ready to use self-healing.

Reviewer: Don't think this sentence adds much value.

> - Cruise control will now keep a look on the cluster regarding if there is any anomaly lurking or not. 

Reviewer: As I said the anomaly detection is always active. It's not enabled through the `self.healing.enabled` property.

By default, the interval at which the detector will run to detect the anomalies is 5 minutes. You can change it by using `anomaly.detection.interval.ms` property.
- If there is an anomaly detected, then the `StrimziNotifier` will create an anomaly resource based on the anomaly detected.
The resource will contain information like anomalyID, anomalyType and anomalyProgress etc. 

> - This `KafkaAnomaly` resource will be watched by both the `operator` and `cruise control.`

Reviewer: I don't get why Cruise Control should watch it and it would be anyway not Cruise Control but the notifier I guess.

> - The operator will then poll the `state` endpoint with `anomaly_detector` substate to know what the current state of the anomaly is.

Reviewer: What if the notifier returns IGNORE to the anomaly detector manager because the self-healing is paused? Should the operator poll the anomaly_detector endpoint anyway? I am not sure about that.

> It will also poll the `state` endpoint with  `executor` state to get the anomalyIds which are currently being fixed

Reviewer: When this polling starts? How polling this endpoint is related to then previous polling?
I mean the operator sends a request on the anomaly_detector endpoint to know if the fix started and then it makes sense to start polling the executor endpoint to know the actual progress.

> - The operator will thenupdate the `anomalyProgress` field of the `KafkaAnomaly` as `FixStarted` if the anomaly fix has been started.

Reviewer: ```suggestion
- The operator will then update the `anomalyProgress` field of the `KafkaAnomaly` as `FixStarted` if the anomaly fix has been started.
```

- To make sure that `anomaly` resource is only deleted when that particular anomaly is fixed, we will again poll the `state` endpoint with  `executor` substate to see if the task are now removed from there.
If there is no task in progress then it would mean that the previous have been completed and we will update the `anomalyprogress` accordingly

> - The `Notifier` will keep a watch on the `KafkaAnomaly` resource, in case it see that `anomalyProgess` has moved to `FixCompleted` then it will that anomaly resource w.r.t the anomalyID which is fixed.

Reviewer 2: ```suggestion
- The `Notifier` will keep a watch on the `KafkaAnomaly` resource, in case it see that `anomalyProgess` has moved to `FixCompleted` then it will remove that anomaly resource.
```

Reviewer 2: I am still not sure about the Notifer removing CRs? The operator will get the info from CC, can't it do the removal? One less step?

- In case a failure is detected and it is not fixable .....<TBD>

#### Pausing the self-healing feature

The user will also get the flexibility to pause the self-healing feature in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the self-healing itself.

> For this we will be having an annotation called `strimzi.io/pause-self-healing-feature`.

Reviewer: I am for `strimzi.io/self-healing-paused` and you should clarify that when self-healing is enabled, it's NOT paused by default so the missing annotation means "false".

> You can set this annotation on the Kafka custom resource to pause the self-healing feature.

Reviewer: I would put an example of `Kafka` custom resource with the annotation set.

Here is how you can enable this annotation:
```sh
kubectl annotate kafka my-cluster strimzi.io/pause-self-healing-feature: "true"
```

> This would pause the self-healing feature for the next set of anomalies but the anomalies which are already being healed at that moment wouldn't be stopped. 

Reviewer: You would need to explain what exactly means "pausing" the self-healing feature. Is the anomaly detected anyway and ignored? Is the anomaly not detected?

> ## Affected/not affected projects

Reviewer: You can improve this section by listing better what's going to be impacted.
It's also missing the operator itself ... which is totally not clear how is going to work when you just mention a "Kubernetes watcher" in the proposal.

This change will affect the `api` module also the way we currently generate the cruise control images. The `strimzi-notifier` module should build as a jar and then be included in the cruise control images.
We would also need to include some of  Kuberenetes dependencies in the cruise control image since we would be creating and watching resource through the Strimzi Notifier.
It will also affect the `installation` of the operator since we will be adding some extra `ClusterRole` and `ClusterRoleBindings` which will need to installed as part of the operator installation.

> ## Rejected Alternatives

Reviewer 2: You need to mention the option of just letting users use CC directly and removing the restriction of self-healing. What impact would this have on the CO operations. Why aren't we just doing that?

Reviewer: +1 ... this is another alternative which drives to a no cloud-native UX but still an alternative.
